File: 剑指Offer/59滑动窗口的最大值.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        numsLen = len(nums)
        if numsLen < k:
            return []

        maxNumberInWindows = []
        maxNumbers = []
        index = 0;

        while index < numsLen:
            if index >= k:
                if nums[index - k] == maxNumbers[0]:
                    maxNumbers.pop(0)

            while len(maxNumbers) != 0 and nums[index] > maxNumbers[len(maxNumbers) - 1]:
                maxNumbers.pop()
            maxNumbers.append(nums[index])

            if index >= k - 1:
                maxNumberInWindows.append(maxNumbers[0])

            index += 1
        
        return maxNumberInWindows

File: 剑指Offer/59滑动窗口的最大值.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> maxNumberInWindow;
        deque<int> maxNumbers;
        int index = 0;
        int numsLen = nums.size();
        if (numsLen < k) return maxNumberInWindow;

        for (; index < numsLen; index++) {
            if (index >= k) {
                if (nums[index - k] == maxNumbers.front()) {
                    maxNumbers.pop_front();
                }
            }
            while (!maxNumbers.empty() && nums[index] > maxNumbers.back()) {
                maxNumbers.pop_back();
            }
            maxNumbers.push_back(nums[index]);
            if (index >= k - 1) {
                maxNumberInWindow.push_back(maxNumbers.front());
            }
        }
        return maxNumberInWindow;
    }
};